cli/internal/cmd/upgradecheck.go

```diff
@@ -0,0 +1,527 @@
+	}
+	log.Debugf("Compatible image measurements are %v", upgrades)
+
+	if len(upgrades) == 0 {
+		return nil, nil
+	}
+
+	return upgrades, nil
+}
+
+func (v *versionCollector) currentVersions(ctx context.Context, log debugLog) (serviceVersion string, imageVersion string, k8sVersion string, err error) {
+	helmClient, err := helm.NewClient(kubectl.New(), constants.AdminConfFilename, constants.HelmNamespace, log)
+	if err != nil {
+		return "", "", "", fmt.Errorf("setting up helm client: %w", err)
+	}
+
+	serviceVersion, err = helmClient.Versions()
+	if err != nil {
+		return "", "", "", fmt.Errorf("getting service versions: %w", err)
+	}
+
+	imageVersion, err = getCurrentImageVersion(ctx, v.checker)
+	if err != nil {
+		return "", "", "", fmt.Errorf("getting image version: %w", err)
+	}
+
+	k8sVersion, err = getCurrentKubernetesVersion(ctx, v.checker)
+	if err != nil {
+		return "", "", "", fmt.Errorf("getting image version: %w", err)
+	}
+
+	return serviceVersion, imageVersion, k8sVersion, nil
+}
+
+// supportedVersions returns slices of supported versions.
+func (v *versionCollector) supportedVersions(cmd *cobra.Command, version string, csp cloudprovider.Provider, log debugLog) (serviceVersion string, imageVersions []versionsapi.Version, k8sVersions []string, err error) {
+	k8sVersions = versions.SupportedK8sVersions()
+	serviceVersion, err = helm.AvailableServiceVersions()
+	if err != nil {
+		return "", nil, nil, fmt.Errorf("loading service versions: %w", err)
+	}
+	imageVersions, err = v.newImages(cmd, version, csp, log)
+	if err != nil {
+		return "", nil, nil, fmt.Errorf("loading image versions: %w", err)
+	}
+
+	return serviceVersion, imageVersions, k8sVersions, nil
+}
+
+func (v *versionCollector) newImages(cmd *cobra.Command, version string, csp cloudprovider.Provider, log debugLog) ([]versionsapi.Version, error) {
+	// find compatible images
+	// image updates should always be possible for the current minor version of the cluster
+	// (e.g. 0.1.0 -> 0.1.1, 0.1.2, 0.1.3, etc.)
+	// additionally, we allow updates to the next minor version (e.g. 0.1.0 -> 0.2.0)
+	// if the CLI minor version is newer than the cluster minor version
+	currentImageMinorVer := semver.MajorMinor(version)
+	currentCLIMinorVer := semver.MajorMinor(v.cliVersion)
+	nextImageMinorVer, err := compatibility.NextMinorVersion(currentImageMinorVer)
+	if err != nil {
+		return nil, fmt.Errorf("calculating next image minor version: %w", err)
+	}
+	log.Debugf("Current image minor version is %s", currentImageMinorVer)
+	log.Debugf("Current CLI minor version is %s", currentCLIMinorVer)
+	log.Debugf("Next image minor version is %s", nextImageMinorVer)
+
+	cliImageCompare := semver.Compare(currentCLIMinorVer, currentImageMinorVer)
+
+	allowedMinorVersions := []string{currentImageMinorVer, nextImageMinorVer}
+	switch {
+	case cliImageCompare < 0:
+		cmd.PrintErrln("Warning: CLI version is older than cluster image version. This is not supported.")
```

Comment:
The version compared here is the one coming from the cluster. That means it is not caught by the config validator.

I agree that we should error in this case. However, in case the force flag is set we need to continue. I left the warning for that case but can also remove it altogether if you think someone using `--force` should not see warnings. We need to continue in case `force` is set as disabling the config validator without disabling this check is useless.